File: src/eliminateCandidate.ml
open Smap
open Sudoku
open Puzzlemap
open hints

let eliminateCandidateApply (p : puzzleMap) (candidate : candidate) (current : current) : current = 

    let update (cell : cell) : cellContents = 
        let cellContents = SMap.get current cell in
        match cellContents with
        | BigNumber _ -> cellContents
        | PencilMarks candidates -> 
            if candidate.cell = cell then PencilMarks(Digits.remove candidate.digit candidates)
            else cellContents
        in

    SMap.ofLookup<cell, cellContents> p.cells update

let eliminateCandidateHintDescription (p: puzzleMap) (candidate : candidate) : hintDescription =
    let cr = makeCandidateReduction (candidate.cell) (Digits.singleton candidate.digit) in

    { hintDescription.primaryHouses = Houses.empty;
      secondaryHouses = Houses.empty;
      candidateReductions = CandidateReductions.singleton cr;
      setCellValueAction = None;
      pointers = CandidateReductions.empty;
      focus = Digits.empty }

let eliminateCandidateStep (p : puzzleMap) (candidate : candidate) (solution : solution) : solution =
    { solution with current = eliminateCandidateApply p candidate solution.current;
                    steps = (Eliminate candidate) :: solution.steps }

File: Sudoku.Lib/Core/eliminateCandidate.fs
module Sudoku.Lib.core.EliminateCandidate

open Sudoku.Lib.compat.Sset
open Sudoku.Lib.compat.oset
open Sudoku.Lib.compat.smap

open Sudoku.Lib.core.Sudoku

let apply (p : Puzzlemap.puzzleMap) (candidate : candidate) (current : current) : current = 

    let update (cell : cell) : cellContents = 
        let cellContents = SMap.get cell current
        match cellContents with
        | BigNumber _ -> cellContents
        | PencilMarks candidates -> 
            if Cell.setElemCompare candidate.cell cell = EQ then PencilMarks(OSet.remove candidate.digit candidates)
            else cellContents

    SMap.ofLookup update p.cells

let description (p: Puzzlemap.puzzleMap) (candidate : candidate) : Hint.description =
    let cr = CandidateReduction.make (candidate.cell) (OSet.singleton candidate.digit) in

    { primaryHouses = OSet.empty();
      secondaryHouses = OSet.empty();
      candidateReductions = [cr];
      setCellValueAction = None;
      pointers = [];
      focus = OSet.empty() }

let step (p : Puzzlemap.puzzleMap) (candidate : candidate) (solution : solution) : solution =
    { solution with current = apply p candidate solution.current;
                    steps = (Eliminate candidate) :: solution.steps }
